### PR TITLE
add maxout layer, including interface and unittest

### PR DESCRIPTION
**Pull request by luotao1**
supported by Su ZhiZhong

### CODE REVIEW

File: paddle/math/Matrix.cpp [hunk 1]
```diff
@@ -26,6 +26,7 @@ limitations under the License. */
 #include "hl_gpu.h"
 #include "hl_table_apply.h"
 #include "hl_top_k.h"
+#include "hl_cnn.h"
```
Comment (qingqing01) [inline paddle/math/Matrix.cpp]: No need to add hl_cnn.h, it is included in hl_gpu.h.

File: paddle/gserver/layers/MaxOutLayer.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  /* Initialize the basic parent class */
+  Layer::init(layerMap, parameterMap);
+
+  /* the size of inputs for maxout-layer is 1 */
+  CHECK_EQ(config_.inputs_size(), 1);
+
+  const MaxOutConfig& conf = config_.inputs(0).maxout_conf();
+  groups_ = conf.groups();
+  channels_ = conf.channels();
+  outputChannels_ = channels_ / groups_;
```
Comment (qingqing01) [inline paddle/gserver/layers/MaxOutLayer.cpp]: Need to check CHECK_EQ(channels%groups_, 0);

File: python/paddle/trainer_config_helpers/layers.py [hunk 3]
```diff
@@ -3334,6 +3335,60 @@ def block_expand_layer(input,
...
+                 groups,
+                 size_x=None,
+                 size_y=None,
+                 name=None,
+                 layer_attr=None):
+    """
+    A layer to do max out on conv layer output.
+      - Input: output of a conv layer.
+      - Output: feature map size same as input. Channel is (input channel) / groups.
+
```
Comment (zzsu) [inline python/paddle/trainer_config_helpers/layers.py]: would be better if we add the reference paper in annotation. 
http://www.jmlr.org/proceedings/papers/v28/goodfellow13.pdf
https://arxiv.org/pdf/1312.6082v4.pdf

File: python/paddle/trainer/config_parser.py [hunk 4]
```diff
@@ -1705,6 +1722,22 @@ def __init__(
             self.set_layer_size(block_expand_conf.block_x * block_expand_conf.block_y
                 * block_expand_conf.channels)
 
+@config_layer('maxout')
+class MaxOutLayer(LayerBase):
+    def __init__(
+            self,
+            name,
+            inputs,
+            device=None):
```
Comment (qingqing01) [inline python/paddle/trainer/config_parser.py]: Better to use **args instead of `device`.

File: paddle/math/Matrix.h [hunk 5]
```diff
@@ -493,16 +493,40 @@ class Matrix : public BaseMatrix {
...
+   *
+   * The row ids and values of these elements are stored in
+   * maxIds and max respectively. where k is the size of maxIds.
+   * And note that the top k elements are not sorted.
+   */
+  virtual void colMax(IVector& maxIds, Matrix& maxVal) {
+    LOG(FATAL) << "not implemented";
+  }
+
+  virtual void maxoutForward(Matrix& a, IVector& id, size_t channels,
```
Comment (zzsu) [inline paddle/math/Matrix.h]: Not sure but is it necessary to add these two functions in class Matrix

File: paddle/gserver/layers/MaxOutLayer.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+#include "MaxOutLayer.h"
+#include "hl_gpu.h"
+#include "hl_cnn.h"
+
+namespace paddle {
+
+REGISTER_LAYER(maxout, MaxOutLayer);
+
+size_t MaxOutLayer::getSize() {
+  CHECK_EQ(inputLayers_.size(), 1UL);
```
Comment (qingqing01) [inline paddle/gserver/layers/MaxOutLayer.cpp]: Remove ` CHECK_EQ(inputLayers_.size(), 1UL);`, it has been add in `init()`.

File: python/paddle/trainer/config_parser.py [hunk 7]
```diff
@@ -1705,6 +1722,22 @@ def __init__(
...
+        super(MaxOutLayer, self).__init__(name, 'maxout', 0, inputs=inputs, device=device)
+        for input_index in xrange(len(self.inputs)):
```
Comment (qingqing01) [inline python/paddle/trainer/config_parser.py]: Only one input layer is checked in `MaxoutLayer.cpp`, so, remove `for` loop.

File: python/paddle/trainer/config_parser.py [hunk 8]
```diff
@@ -1705,6 +1722,22 @@ def __init__(
             self.set_layer_size(block_expand_conf.block_x * block_expand_conf.block_y
                 * block_expand_conf.channels)
```
Comment (zzsu) [inline python/paddle/trainer/config_parser.py]: maxout is like an activation, better check the activation of the conv before it is linear activated.

File: python/paddle/trainer_config_helpers/layers.py (hunk 3 above)
Comment (luotao1) [inline python/paddle/trainer_config_helpers/layers.py]: add the reference paper

File: paddle/math/Matrix.h (hunk 5 above)
Comment (luotao1) [inline paddle/math/Matrix.h]: @qingqing01 suggest me to package these two functions in class Matrix, thus, we can do unittest in test_matrixCompare

File: paddle/math/Matrix.cpp (hunk 1 above)
Comment (luotao1) [inline paddle/math/Matrix.cpp]: remove it

File: paddle/gserver/layers/MaxOutLayer.cpp (hunk 6 above)
Comment (luotao1) [inline paddle/gserver/layers/MaxOutLayer.cpp]: remove it

File: paddle/gserver/layers/MaxOutLayer.cpp (hunk 2 above)
Comment (luotao1) [inline paddle/gserver/layers/MaxOutLayer.cpp]: add check

File: python/paddle/trainer/config_parser.py (hunk 4 above)
Comment (luotao1) [inline python/paddle/trainer/config_parser.py]: use `**xargs`

File: python/paddle/trainer/config_parser.py (hunk 7 above)
Comment (luotao1) [inline python/paddle/trainer/config_parser.py]: remove the loop

File: python/paddle/trainer/config_parser.py (hunk 8 above)
Comment (luotao1) [inline python/paddle/trainer/config_parser.py]: add check linear activation

File: python/paddle/trainer_config_helpers/layers.py [hunk 9]
```diff
@@ -3334,6 +3335,60 @@ def block_expand_layer(input,
 
 
 @wrap_name_default()
+@layer_support()
+def maxout_layer(input,
+                 channels,
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: channels=None,  channels can be got from the previous layer. So, better to automatically setting.

**Pull request merged**